components/raftstore/src/store/async_io.rs

```diff
@@ -0,0 +1,949 @@
+    pub cut_logs: Option<(u64, u64)>,
+    pub raft_state: Option<RaftLocalState>,
+    pub messages: Vec<RaftMessage>,
+    pub proposal_times: Vec<Instant>,
+}
+
+impl<EK, ER> AsyncWriteTask<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    pub fn new(region_id: u64, peer_id: u64, ready_number: u64) -> Self {
+        Self {
+            region_id,
+            peer_id,
+            ready_number,
+            send_time: Instant::now(),
+            kv_wb: None,
+            raft_wb: None,
+            entries: vec![],
+            cut_logs: None,
+            raft_state: None,
+            messages: vec![],
+            proposal_times: vec![],
+        }
+    }
+
+    pub fn has_data(&self) -> bool {
+        !(self.raft_state.is_none()
+            && self.entries.is_empty()
+            && self.cut_logs.is_none()
+            && self.kv_wb.as_ref().map_or(true, |wb| wb.is_empty())
+            && self.raft_wb.as_ref().map_or(true, |wb| wb.is_empty()))
+    }
+
+    /// Sanity check for robustness.
+    pub fn valid(&self) -> Result<()> {
+        if self.region_id == 0 || self.peer_id == 0 || self.ready_number == 0 {
+            return Err(box_err!(
+                "invalid id, region_id {}, peer_id {}, ready_number {}",
+                self.region_id,
+                self.peer_id,
+                self.ready_number
+            ));
+        }
+        let last_index = self.entries.last().map_or(0, |e| e.get_index());
+        if let Some((from, _)) = self.cut_logs {
+            if from <= last_index {
+                // Entries are put and deleted in the same writebatch.
+                return Err(box_err!(
+                    "invalid cut logs, last_index {}, cut_logs {:?}",
+                    last_index,
+                    self.cut_logs
+                ));
+            }
+        }
+        if last_index != 0
+            && self
+                .raft_state
+                .as_ref()
+                .map_or(true, |r| r.get_last_index() != last_index)
+        {
+            return Err(box_err!(
+                "invalid raft state, last_index {}, raft_state {:?}",
+                last_index,
+                self.raft_state
+            ));
+        }
+        Ok(())
+    }
+}
+
+/// Message that can be sent to AsyncWriteWorker
+pub enum AsyncWriteMsg<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    WriteTask(AsyncWriteTask<EK, ER>),
+    Shutdown,
+}
+
+/// AsyncWriteBatch is used for combining several AsyncWriteTask into one.
+struct AsyncWriteBatch<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    pub kv_wb: EK::WriteBatch,
+    pub raft_wb: ER::LogBatch,
+    // region_id -> (peer_id, ready_number)
+    pub readies: HashMap<u64, (u64, u64)>,
+    // Write raft state once for a region everytime writing to disk
+    pub raft_states: HashMap<u64, RaftLocalState>,
+    pub state_size: usize,
+    pub tasks: Vec<AsyncWriteTask<EK, ER>>,
+    _phantom: PhantomData<EK>,
```

Comment:
Is it necessary? `tasks` already has the generic arguments.

Reply:
Good catch. These codes here have changed many times. I forget to remove it.